packages/utils/src/ast-utils/helpers.ts

```diff
@@ -1,38 +1,40 @@
 import { AST_NODE_TYPES, AST_TOKEN_TYPES, TSESTree } from '../ts-estree';
 
-type ObjectEntry<BaseType> = [keyof BaseType, BaseType[keyof BaseType]];
+type ObjectEntry<BaseType> = BaseType extends unknown
+  ? [keyof BaseType, BaseType[keyof BaseType]]
+  : never;
```

Comment:
I do agree that distributing it here makes more sense—but FYI, the code compiles without this distribution. Even Intellisense is taking forever on my device because of the huge types here, so I can't debug deeper about why.

Reply:
I'm not sure how much time it took before but it's so slow on my Air and Pro, so it takes ages to check things

Not sure if I asked for this change and introduced another performance issue here

Reply:
It was _relatively_ faster? I checked out main and at least intellisense seems reasonable. I never thought `Extract` would be that big of an issue, but thinking about it, it makes sense, since intersections can take the fast path when checking assignability.

Reply:
Yeah, also didn't think of it as a big issue

Reply:
I checked that we can save result of `Extract` as generic type and can reuse them – https://github.com/typescript-eslint/typescript-eslint/pull/4545#discussion_r895631463

It can improve IntelliSense slightly, still checking it

Reply:
Can you quickly check the PR again? It looks more responsive for me

Reply:
Aye, could confirm it works much much better 👍

Reply:
Thank you for checking!